pkg/gitrepository/manager.go

```diff
@@ -0,0 +1,140 @@
+package gitrepository
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	profilesv1 "github.com/weaveworks/profiles/api/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//Manager is responsible for managing gitrepository resources
+type Manager struct {
+	kClient   Kubernetes
+	namespace string
+	ctx       context.Context
+	timeout   time.Duration
+	interval  time.Duration
+}
+
+//Instance contains a tag and path of profile.yaml
+type Instance struct {
+	Tag  string
+	Path string
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate -o fakes/fake_kubernetes.go . Kubernetes
+//Kubernetes interface for interacting with Kubernetes
+type Kubernetes interface {
+	Get(ctx context.Context, key client.ObjectKey, obj client.Object) error
+	Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error
+	Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error
+}
+
+//NewManager returns a Manager struct
+func NewManager(ctx context.Context, namespace string, kClient Kubernetes, timeout, interval time.Duration) *Manager {
+	return &Manager{
+		kClient:   kClient,
+		namespace: namespace,
+		ctx:       ctx,
+		timeout:   timeout,
+		interval:  interval,
+	}
+}
+
+//CreateAndWaitForResources creates the gitrepository resources and waits for them to be created
+func (m *Manager) CreateAndWaitForResources(r profilesv1.Repository, instances []Instance) ([]*sourcev1.GitRepository, error) {
+	var gitResources []*sourcev1.GitRepository
+	for _, instance := range instances {
+		gitRes := makeGitRepository(r, instance.Tag, instance.Path, m.namespace)
+		err := m.kClient.Create(m.ctx, gitRes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create gitrepository: %w", err)
+		}
+		gitResources = append(gitResources, gitRes)
+	}
+
+	for _, gitRes := range gitResources {
+		err := m.waitForURL(gitRes)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return gitResources, nil
+}
+
+func (m *Manager) waitForURL(gitRes *sourcev1.GitRepository) error {
+	ctx, cancel := context.WithTimeout(context.Background(), m.timeout)
+	defer cancel()
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("timed out waiting for %s/%s gitrepository.Status.URL to be populated", gitRes.Namespace, gitRes.Name)
+		case <-time.After(m.interval):
+			err := m.kClient.Get(m.ctx, client.ObjectKeyFromObject(gitRes), gitRes)
```

Comment:
You could even go crazy and ignore the this case in here, and put it in the beginning of the for loop then leave this empty. Because it's literally only there to progress the for loop. So like this:

```go
	for {
		select {
		case <-ctx.Done():
			return fmt.Errorf("timed out waiting for %s/%s gitrepository.Status.URL to be populated", gitRes.Namespace, gitRes.Name)
		case <-time.After(m.interval):
		}      
		if err := m.kClient.Get(m.ctx, client.ObjectKeyFromObject(gitRes), gitRes); err != nil {
			return fmt.Errorf("failed to get gitrepository: %w", err)
		}
		if gitRes.Status.URL != "" {
			return nil
		}
	}
```

Reply:
I prefer it as is :smile:

Reply:
Okay. :D